entities/component.py

```diff
@@ -172,12 +173,17 @@ def __init__(self, uuid: str, name: Name, description: Description,
         self.version = version
         self.created = created
         self.deprecated = deprecated
-        self.category = category
+        self.generated_by = generated_by
+        self.categories = categories
         self.schematic_only = schematic_only
         self.default_value = default_value
         self.prefix = prefix
         self.signals = []  # type: List[Signal]
         self.variants = []  # type: List[Variant]
+        self.approvals = []  # type: List[str]
+
+    def add_approval(self, approval: str) -> None:
+        self.approvals.append(approval)
```

Comment:
Shouldn't the approval be an entity in itself? See next comment.

Reply:
That's (almost) not possible because the data structure of approvals depends on the type of approval, and there are several dozens of different approval types. In LibrePCB, an approval is therefore just of type `SExpression` but we don't have such a type in this generator since there is no abstraction for S-Expression nodes.

Reply:
Hmm. I'm not sure I understand. But we could just define it something like this:
```python
class Approval:
    """Helper class to represent a single named string value"""
    def __init__(self, value: Union[str,whatever, ...]):
        self.value = value

    def __str__(self) -> str:
        return '(approval {})'.format(self.name, self.value)
```

Then whatever we pass in as the `value` will now how to serialize itself.

But I'm probably misunderstanding something?

Reply:
I'm not sure if your code works. But ok I agree at least it would work when defining `Approval` as an alias for `str` since working with `str` directly works fine.

Anyway, I think for long term we need to have a Python wrapper for the C++ classes since it is very cumbersome, time consuming and error-prone to implement the whole file format again. So I wouldn't "vergolden" the entity classes :slightly_smiling_face:

Btw, I don't get the reason for defining separate classes for every *instance* of a property. For example:

```python
class Fill:
    def __init__(self, value: bool):
        self.value = value

    def __str__(self) -> str:
        return '(fill {})'.format(self.value.lower())

class GrabArea:
    def __init__(self, value: bool):
        self.value = value

    def __str__(self) -> str:
        return '(grab_area {})'.format(self.value.lower())

class Polygon():
    def __init__(fill: Fill, grab_area: GrabArea):
        self.fill = fill
        self.grab_area = grab_area

    def __str__(self) -> str:
        return '(polygon {} {})'.format(self.fill, self.grab_area)
```

Instead of having the type only once, like this:

```python
class BoolValue:
    def __init__(self, value: bool):
        self.value = value

    def __str__(self) -> str:
        return str(self.value).lower()

class Polygon():
    def __init__(fill: BoolValue, grab_area: BoolValue):
        self.fill = fill
        self.grab_area = grab_area

    def __str__(self) -> str:
        return '(polygon (fill {}) (grab_area {}))'.format(self.fill, self.grab_area)
```

Or one could consider going even further, allowing to pass Python native types where available:

```python
class BoolValue:
    def __init__(self, value: bool):
        self.value = value

    def __str__(self) -> str:
        return str(self.value).lower()

class Polygon():
    def __init__(fill: bool, grab_area: bool):
        self.fill = fill
        self.grab_area = grab_area

    def __str__(self) -> str:
        return '(polygon (fill {}) (grab_area {}))'.format(BoolValue(self.fill), BoolValue(self.grab_area))
```

Wouldn't that be more Pythonic? At least it would make the API easier to use :slightly_smiling_face: But maybe there's a specific reason for the current design?

Reply:
>Wouldn't that be more Pythonic? At least it would make the API easier to use 🙂 But maybe there's a specific reason for the current design?

The reason was to have type-checked named parameters, so that it is clear when instantiating what the value is. So instead of having weakly typed strings, we have strongly typed names. So:

```python
Polygon(Fill(true), GrabAre(false)
```
instead of
```python
Polygon(BoolValue(true), BoolValue(false))
```

Reply:
I agree with @rnestler, I think there is value in having typed properties (in combination with mypy).

However, we can do better to avoid repetition. For example:

```python
class BoolValue:
    def __init__(self, value: bool, name: str):
        self.value = value
        self.name = name

    def __str__(self) -> str:
        return f'({self.name} {self.value.lower()})'
    
class Fill(BoolValue):
    def __init__(self, value: bool):
        super().__init__(value, 'fill')

class GrabArea(BoolValue):
    def __init__(self, value: bool):
        super().__init__(value, 'grab_area')
```

Or we could even go a step further and write a class decorator or a metaclass that automatically generates the implementation from some kind of DSL. Advantage is that we have less repetitive code in our entity definitions. Downside is that IDE support / autocompletion gets worse, because some implementations are generated dynamically with metaprogramming. I'm not sure if this is a good tradeoff.

Reply:
Okay, but since we usually pass the arguments with kwargs (`fill=Fill(True)`) it feels a bit redundant to me. Also the extreme type safety is actually not needed since everything generated by these scripts will be validated much more strictly by `librepcb-cli` (on CI of libraries). IMHO the KISS principle could therefore be helpful in reducing the amount of code and speeding up development without introducing any relevant risk.

But I think in the end it's a matter of opinions so the discussion won't help :slightly_smiling_face: